tests/components/tradfri/test_cover.py

```diff
@@ -0,0 +1,219 @@
+    """Mock a tradfri cover/blind."""
+    if test_features is None:
+        test_features = {}
+    if test_state is None:
+        test_state = {}
+    mock_cover_data = Mock(**test_state)
+
+    dev_info_mock = MagicMock()
+    dev_info_mock.manufacturer = "manufacturer"
+    dev_info_mock.model_number = "model"
+    dev_info_mock.firmware_version = "1.2.3"
+    _mock_cover = Mock(
+        id=f"mock-cover-id-{device_number}",
+        reachable=True,
+        observe=Mock(),
+        device_info=dev_info_mock,
+        has_light_control=False,
+        has_socket_control=False,
+        has_blind_control=True,
+        has_signal_repeater_control=False,
+        has_air_purifier_control=False,
+    )
+    _mock_cover.name = f"tradfri_cover_{device_number}"
+
+    # Set supported features for the covers.
+    blind_control = BlindControl(_mock_cover)
+
+    # Store the initial state.
+    setattr(blind_control, "blinds", [mock_cover_data])
+    _mock_cover.blind_control = blind_control
+    return _mock_cover
+
+
+async def test_cover(hass, mock_gateway, mock_api_factory):
+    """Test that covers are correctly added."""
+    state = {
+        "current_cover_position": 40,
+    }
+
+    mock_gateway.mock_devices.append(mock_cover(test_state=state))
+    await setup_integration(hass)
+
+    cover_1 = hass.states.get("cover.tradfri_cover_0")
+    assert cover_1 is not None
+    assert cover_1.state == "open"
+    assert cover_1.attributes["current_position"] == 60
+
+
+async def test_cover_observed(hass, mock_gateway, mock_api_factory):
+    """Test that covers are correctly observed."""
+    state = {
+        "current_cover_position": 1,
+    }
+
+    cover = mock_cover(test_state=state)
+    mock_gateway.mock_devices.append(cover)
+    await setup_integration(hass)
+    assert len(cover.observe.mock_calls) > 0
+
+
+async def test_cover_available(hass, mock_gateway, mock_api_factory):
+    """Test cover available property."""
+
+    cover = mock_cover(test_state={"current_cover_position": 1}, device_number=1)
+    cover.reachable = True
+
+    cover2 = mock_cover(test_state={"current_cover_position": 1}, device_number=2)
+    cover2.reachable = False
+
+    mock_gateway.mock_devices.append(cover)
+    mock_gateway.mock_devices.append(cover2)
+    await setup_integration(hass)
+
+    assert hass.states.get("cover.tradfri_cover_1").state == "open"
+    assert hass.states.get("cover.tradfri_cover_2").state == "unavailable"
+
+
+def create_all_set_position_on_cases():
+    """Create all test cases."""
+    all_test_cases = [
+        ["test_features", "test_data", "expected_result", "device_id"],
+        [],
+    ]
+    index = 1
+    for test_case in SET_POSITION_TEST_CASES:
+        case = deepcopy(test_case)
+        case.append(index)
+        index += 1
+        all_test_cases[1].append(case)
+
+    return all_test_cases
+
+
+@pytest.mark.parametrize(*create_all_set_position_on_cases())
```

Comment:
this seems a complicated way to parametrize the test.
Could it not be done with
`@pytest.mark.parametrize("test_features, test_data, expected_result, device_id",[[a1, a2, a3, a4], [b1, b2, b3, b4]])`
?

Reply:
I agree, figured there was a reason it was there in the light platform tests in the first place and that the reason was a matter of principle in the HA project. I'll change it.